Modules/Scripted/WebServer/WebServer.py

```diff
@@ -0,0 +1,1507 @@
+          studyDataset.StudyInstanceUID = dataset.StudyInstanceUID
+          studyDataset.AccessionNumber = dataset.AccessionNumber
+          studyDataset.InstanceAvailability = u'ONLINE'
+          studyDataset.ModalitiesInStudy = list(modalitiesInStudy)
+          studyDataset.ReferringPhysicianName = dataset.ReferringPhysicianName
+          studyDataset[self.retrieveURLTag] = pydicom.dataelem.DataElement(
+              0x00080190, "UR", "TODO: provide WADO-RS RetrieveURL")
+          studyDataset.PatientName = dataset.PatientName
+          studyDataset.PatientID = dataset.PatientID
+          studyDataset.PatientBirthDate = dataset.PatientBirthDate
+          studyDataset.PatientSex = dataset.PatientSex
+          studyDataset.StudyID = dataset.StudyID
+          studyDataset[self.numberOfStudyRelatedSeriesTag] = pydicom.dataelem.DataElement(
+              self.numberOfStudyRelatedSeriesTag, "IS", str(numberOfStudyRelatedSeries))
+          studyDataset[self.numberOfStudyRelatedInstancesTag] = pydicom.dataelem.DataElement(
+              self.numberOfStudyRelatedInstancesTag, "IS", str(numberOfStudyRelatedInstances))
+          jsonDataset = studyDataset.to_json(studyDataset)
+          studyResponseString += jsonDataset.encode() + b","
+      if studyResponseString.endswith(b','):
+        studyResponseString = studyResponseString[:-1]
+      studyResponseString += b']'
+      responseBody = studyResponseString
+    elif splitPath[4] == b'metadata':
+      self.logMessage('returning metadata')
+      contentType = b'application/json'
+      responseBody = b"["
+      studyUID = splitPath[3].decode()
+      series = slicer.dicomDatabase.seriesForStudy(studyUID)
+      for serie in series:
+        seriesInstances = slicer.dicomDatabase.instancesForSeries(serie)
+        for instance in seriesInstances:
+          dataset = pydicom.dcmread(slicer.dicomDatabase.fileForInstance(instance), stop_before_pixels=True)
+          jsonDataset = dataset.to_json()
+          responseBody += jsonDataset.encode() + b","
+      if responseBody.endswith(b','):
+        responseBody = responseBody[:-1]
+      responseBody += b']'
+    return contentType, responseBody
+
+  def handleWADOURI(self, parsedURL, requestBody):
+    """
+    Handle wado uri by returning the binary part10 contents of the dicom file
+    :param parsedURL: the REST path and arguments
+    :param requestBody: the binary that came with the request
+    """
+    q = urlparse.parse_qs(parsedURL.query)
+    try:
+      instanceUID = q[b'objectUID'][0].decode().strip()
+    except KeyError:
+      return None,None
+    self.logMessage('found uid %s' % instanceUID)
+    contentType = b'application/dicom'
+    path = slicer.dicomDatabase.fileForInstance(instanceUID)
+    fp = open(path, 'rb')
+    responseBody = fp.read()
+    fp.close()
+    return contentType, responseBody
+
+
+#
+# SlicerRequestHandler
+#
+
+class SlicerRequestHandler(object):
```

Comment:
Could we move out each request handler to a separate file?
It would make this file shorter and simpler and it could server as example for extensions defining and registering their own custom request handlers.